nemo/collections/tts/modules/audio_codec_modules.py

```diff
@@ -698,3 +716,429 @@ def decode(self, indices: torch.Tensor, input_len: torch.Tensor) -> torch.Tensor
+                kernel_sizes=resblock_kernel_sizes,
+                dilations=resblock_dilation_sizes,
+                activation=activation,
+            )
+            self.res_layers.append(res_layer)
+
+        self.post_activation = CodecActivation(activation, channels=in_channels)
+        self.post_conv = Conv1dNorm(in_channels=in_channels, out_channels=1, kernel_size=out_kernel_size)
+        self.out_activation = nn.Tanh()
+
+    @property
+    def input_types(self):
+        return {
+            "inputs": NeuralType(('B', 'D', 'T_encoded'), VoidType()),
+            "input_len": NeuralType(tuple('B'), LengthsType()),
+        }
+
+    @property
+    def output_types(self):
+        return {
+            "audio": NeuralType(('B', 'T_audio'), AudioSignal()),
+            "audio_len": NeuralType(tuple('B'), LengthsType()),
+        }
+
+    def remove_weight_norm(self):
+        self.pre_conv.remove_weight_norm()
+        for up_sample_conv in self.up_sample_conv_layers:
+            up_sample_conv.remove_weight_norm()
+        for res_layer in self.res_layers:
+            res_layer.remove_weight_norm()
+
+    @typecheck()
+    def forward(self, inputs, input_len):
+        audio_len = input_len
+        # [B, C, T_encoded]
+        out = self.pre_conv(inputs=inputs, input_len=audio_len)
+        for act, res_layer, up_sample_conv, up_sample_rate in zip(
+            self.activations, self.res_layers, self.up_sample_conv_layers, self.up_sample_rates
+        ):
+            audio_len = audio_len * up_sample_rate
+            out = act(out)
+            # [B, C / 2, T * up_sample_rate]
+            out = up_sample_conv(inputs=out, input_len=audio_len)
+            out = res_layer(inputs=out, input_len=audio_len)
+
+        out = self.post_activation(out)
+        # [B, 1, T_audio]
+        out = self.post_conv(inputs=out, input_len=audio_len)
+        audio = self.out_activation(out)
+        audio = rearrange(audio, "B 1 T -> B T")
+        return audio, audio_len
+
+
+class MelSpectrogramProcessor(NeuralModule):
+    """
+    Wrapper interface for computing mel spectrogram for codec training.
+    """
+
+    def __init__(self, sample_rate: int, win_length: int, hop_length: int, mel_dim: int = 80, log_guard: float = 1.0):
+        super(MelSpectrogramProcessor, self).__init__()
+        self.mel_dim = mel_dim
+        self.hop_length = hop_length
+        self.preprocessor = AudioToMelSpectrogramPreprocessor(
+            sample_rate=sample_rate,
+            highfreq=None,
+            features=mel_dim,
+            pad_to=1,
+            exact_pad=True,
+            n_window_size=win_length,
+            n_window_stride=hop_length,
+            window_size=False,
+            window_stride=False,
+            n_fft=win_length,
+            mag_power=1.0,
+            log=True,
+            log_zero_guard_type="add",
+            log_zero_guard_value=log_guard,
+            mel_norm=None,
+            normalize=None,
```

Comment:
curious: Are there any effects on the performance of encoder with normalization?

Reply:
If you mean the `normalize` parameter which if I understand correctly does CMVN, then that would probably hurt compression performance because it would erase some information like channel effects.

If you mean `mel_norm`, then that goes to discussions long ago on whether we should use log(mel + 1.0), or the TTS standard definition of log(max(norm(mel), 1E-5)). If I remember correctly the results were mixed/inconclusive when we were using the original EnCodec losses, and after the changes in https://github.com/NVIDIA/NeMo/pull/7468 the version without normalization was generally better.